File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
private:
   unsigned long long f(vector<int>& piles, int mid){
       unsigned long long sum=0;
        for(auto pile : piles){
          sum+=  ceil((double)pile/mid);
        }
        return sum;
    }
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        if(h<piles.size()) return 0;

        int low=1,high=1e9;
        int ans=0;

        while(low<=high){
            int mid=(low+high)/2;
         unsigned   long long required=f(piles,mid);
            if(required<=h){
                high=mid-1;
                ans=mid;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        

        maxi=max(piles)

        mini=1
        # if mini > h:
        #     return -1
        
        def cal(piles,mid):
            sum=0
            for pile in piles:
              sum=sum +  ceil(pile/mid)
            return sum
        ans=0
        while mini<=maxi:
            mid = (mini+maxi) // 2

            required=cal(piles,mid)

            if required <= h:
                ans=mid
                maxi=mid-1
            else:
                mini=mid+1

        return ans    

